atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java

```diff
@@ -0,0 +1,816 @@
+                    }
+                    return results;
+                } finally {
+                    iter.close();
+                }
+            }
+        });
+    }
+
+    private static final int INSERT_PARTITION_COUNT = 1000;
+    private static final long INSERT_PARTITION_BYTES = 2048 * 1024;
+
+    public Function<Entry<Cell, byte[]>, Long> getByteSizingFunction() {
+        return new Function<Entry<Cell, byte[]>, Long>(){
+            @Override
+            public Long apply(Entry<Cell, byte[]> entry) {
+                return Cells.getApproxSizeOfCell(entry.getKey()) + entry.getValue().length;
+            }
+        };
+    }
+
+    public Function<Entry<Cell, Value>, Long> getValueSizingFunction() {
+        return new Function<Entry<Cell, Value>, Long>(){
+            @Override
+            public Long apply(Entry<Cell, Value> entry) {
+                return Cells.getApproxSizeOfCell(entry.getKey()) + entry.getValue().getContents().length;
+            }
+        };
+    }
+
+    private void put(TableReference tableRef, Map<Cell, byte[]> values, long timestamp, boolean idempotent) {
+        final Iterable<List<Entry<Cell, byte[]>>> batches = partitionByCountAndBytes(
+                values.entrySet(), INSERT_PARTITION_COUNT, INSERT_PARTITION_BYTES, tableRef, getByteSizingFunction());
+        runWrite(tableRef, new Function<DbWriteTable, Void>() {
+            @Override
+            public Void apply(DbWriteTable table) {
+                for (List<Entry<Cell, byte[]>> batch : batches) {
+                    try {
+                        table.put(batch, timestamp);
+                    } catch (KeyAlreadyExistsException e) {
+                        if (idempotent) {
+                            putIfNotUpdate(tableRef, table, batch, timestamp, e);
+                        } else {
+                            throw e;
+                        }
+                    }
+                }
+                return null;
+            }
+        });
+    }
+
+    @Override
+    public void put(TableReference tableRef, final Map<Cell, byte[]> values, final long timestamp) throws KeyAlreadyExistsException {
+        put(tableRef, values, timestamp, true);
+    }
+
+    private void putIfNotUpdate(TableReference tableRef, DbWriteTable table, List<Entry<Cell, Value>> batch, KeyAlreadyExistsException e) {
+        Map<Cell, Long> timestampByCell = Maps.newHashMap();
+        for (Entry<Cell, Value> entry : batch) {
+            timestampByCell.put(entry.getKey(), entry.getValue().getTimestamp() + 1);
+        }
+        Map<Cell, Value> results = get(tableRef, timestampByCell);
+
+        ListIterator<Entry<Cell, Value>> iter = batch.listIterator();
+        while (iter.hasNext()) {
+            Entry<Cell, Value> entry = iter.next();
+            Cell key = entry.getKey();
+            Value value = entry.getValue();
+            if (results.containsKey(key)) {
+                if (results.get(key).equals(value)) {
```

Comment:
nit: would be nice to traverse map only once (i.e. `Value result = results.get(key); if (Objects.equals(result, value)) ...`)